cmd/recovereth/main.go

```diff
@@ -0,0 +1,396 @@
+	aesBlock, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	stream := cipher.NewCTR(aesBlock, iv)
+	outText := make([]byte, len(inText))
+	stream.XORKeyStream(outText, inText)
+	return outText, err
+}
+
+// decryptKeyV3x decrypts does not validate MAC during decryption
+func decryptKeyV3x(crypto cryptoJSON, auth string) (keyBytes []byte, err error) {
+	//keyId = uuid.Parse(keyProtected.Id)
+
+	iv, err := hex.DecodeString(crypto.CipherParams.IV)
+	if err != nil {
+		return nil, err
+	}
+
+	cipherText, err := hex.DecodeString(crypto.CipherText)
+	if err != nil {
+		return nil, err
+	}
+
+	derivedKey, err := getKDFKey(crypto, auth)
+	if err != nil {
+		return nil, err
+	}
+
+	plainText, err := aesCTRXOR(derivedKey[:16], cipherText, iv)
+	if err != nil {
+		return nil, err
+	}
+	return plainText, err
+}
+
+func testPasswordPrefixes(keyjson, auth string) error {
+	l := []string{"", "\n", "\r\n", " ", "\r"}
+	k := new(encryptedKeyJSONV3)
+	if err := json.Unmarshal([]byte(keyjson), k); err != nil {
+		log.Fatal(err)
+	}
+	addr := common.HexToAddress(k.Address)
+	for _, prefix := range l {
+		for _, suffix := range l {
+			pw := fmt.Sprintf("%v%v%v", prefix, auth, suffix)
+			keyBytes, err := decryptKeyV3x(k.Crypto, pw)
+			fmt.Printf("auth: %x\n", pw)
+			if err != nil {
+				return err
+			}
+			key := crypto.ToECDSAUnsafe(keyBytes)
+			if err != nil {
+				return err
+			}
+			recovered := crypto.PubkeyToAddress(key.PublicKey)
+			fmt.Printf("found address %s (looking for %s)\n", recovered.Hex(), addr.Hex())
+			if recovered == addr {
+				fmt.Printf("Recovery successfull!\n")
+				fmt.Printf("Recovered password: \n------------\n%v\nHex: 0x%x\n------------\n", pw, pw)
+				fmt.Printf("Please report to the go-ethereum repository "+
+					"that you were able to recover a wallet with prefix 0x%x and suffix 0x%x.", prefix, suffix)
+				return nil
+			}
+		}
+	}
+	return errors.New("Exhausted")
+}
+func testPasswordCrops(keyjson, auth string) error {
+	k := new(encryptedKeyJSONV3)
+	if err := json.Unmarshal([]byte(keyjson), k); err != nil {
+		log.Fatal(err)
+	}
+	addr := common.HexToAddress(k.Address)
+	for i:= 0 ; i < len(auth); i++ {
+		pw := auth[:i]
+		keyBytes, err := decryptKeyV3x(k.Crypto, pw)
+		fmt.Printf("auth: %x\n", pw)
+		if err != nil {
+			return err
+		}
+		key := crypto.ToECDSAUnsafe(keyBytes)
+		if err != nil {
+			return err
+		}
+		recovered := crypto.PubkeyToAddress(key.PublicKey)
+		fmt.Printf("found address %s (looking for %s)\n", recovered.Hex(), addr.Hex())
+		if recovered == addr {
+			fmt.Printf("Recovery successfull!\n")
+			fmt.Printf("Recovered password: \n------------\n%v\nHex: 0x%x\n------------\n", pw, pw)
+			fmt.Printf("Please report to the go-ethereum repository "+
+				"that you were able to recover a wallet by cropping %d-length password to length %d ", len(auth), len(pw))
+			return nil
+		}
+	}
+	return errors.New("Exhausted")
+}
+
+func testRecoveryBitFlip(keyjson, auth string, i uint) error {
```

Comment:
An infinite loop may occur in the testRecoveryBitFlip function if the address is not found. You can add an additional parameter `maxIterations uint` to limit the number of iterations. If the function cannot find the address in the maximum number of iterations, it should return an error.

Here is an example of how I see the implementation of this function:

```go
func testRecoveryBitFlip(keyjson, auth string, i, maxIterations uint) error {
	// i = 0 starts on the Ciphertext
	// i = 256 starts on the IV
	// i = 384 starts on the SALT
	for iter := uint(0); iter < maxIterations; iter++ {
		k := new(encryptedKeyJSONV3)

		if err := json.Unmarshal([]byte(keyjson), k); err != nil {
			return err
		}
		addr := common.HexToAddress(k.Address)

		k2, err := flipBit(k.Crypto, i)
		if err != nil {
			return err
		}

		fmt.Printf("C  %s -> %s \n", k.Crypto.CipherText, k2.CipherText)
		fmt.Printf("IV %s -> %s \n", k.Crypto.CipherParams.IV, k2.CipherParams.IV)
		fmt.Printf("Sa %s -> %s \n", k.Crypto.KDFParams["salt"].(string), k2.KDFParams["salt"].(string))

		keyBytes, err := decryptKeyV3x(k2, auth)
		if err != nil {
			return err
		}
		key := crypto.ToECDSAUnsafe(keyBytes)
		if err != nil {
			return err
		}
		recovered := crypto.PubkeyToAddress(key.PublicKey)
		fmt.Printf("found address %s (looking for %s)\n", recovered.Hex(), addr.Hex())
		if recovered == addr {
			fmt.Printf("Recovery successfull!\n")
			json.Marshal(k2)
			json, _ := json.MarshalIndent(k2, "", "    ")
			fmt.Printf("Recovered json-file: \n------------\n%v\n------------\n", string(json))
			fmt.Printf("Please report to the go-ethereum repository "+
				"that you were able to recover a wallet by 'bitflip'-manipulation at bit %d.", i)
			return nil
		}
	}
	return fmt.Errorf("unable to recover wallet after %d iterations", maxIterations)
}
```